plugins/algoltek-usb/fu-algoltek-usb-firmware.c

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright (C) 2023 Ling.Chen <ling.chen@algoltek.com.tw>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include <fwupdplugin.h>
+
+#include "fu-algoltek-usb-firmware.h"
+#include "fu-algoltek-usb-struct.h"
+
+struct _FuAlgoltekUsbFirmware {
+	FuFirmware parent_instance;
+};
+
+G_DEFINE_TYPE(FuAlgoltekUsbFirmware, fu_algoltek_usb_firmware, FU_TYPE_FIRMWARE)
+
+#define FU_ALGOLTEK_HEADER 0x4B45544C4F474C41 // ALGOLTEK
+
+#define FU_ALGOLTEK_FIRMWARE_SIZE 0x20000
+#define FU_ALGOLTEK_ISP_SIZE	  0x1000
+
+static gboolean
+fu_algoltek_usb_firmware_validate(FuFirmware *firmware, GBytes *fw, gsize offset, GError **error)
+{
+	guint64 magic = 0;
+	/* First byte is header length */
+	offset += 1;
+
+	if (!fu_memread_uint64_safe(g_bytes_get_data(fw, NULL),
+				    g_bytes_get_size(fw),
+				    offset,
+				    &magic,
+				    G_LITTLE_ENDIAN,
+				    error)) {
+		g_prefix_error(error, "failed to read magic: ");
+		return FALSE;
+	}
+
+	if (magic != FU_ALGOLTEK_HEADER) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_INVALID_FILE,
+			    "invalid magic, expected 0x%08lX got 0x%08lX",
+			    (guint64)FU_ALGOLTEK_HEADER,
+			    magic);
+		return FALSE;
+	}
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_algoltek_usb_firmware_parse(FuFirmware *firmware,
+			       GBytes *fw,
+			       gsize offset,
+			       FwupdInstallFlags flags,
+			       GError **error)
+{
+	g_autoptr(FuFirmware) img_ISP = fu_firmware_new();
+	g_autoptr(FuFirmware) img_payload = fu_firmware_new();
+	g_autoptr(GBytes) blob_ISP = NULL;
+	g_autoptr(GBytes) blob_payload = NULL;
+	g_autoptr(GByteArray) headerArray = g_byte_array_new();
+	g_autoptr(GBytes) blob_header = NULL;
+	gchar *version;
+
+	blob_header = fu_bytes_new_offset(fw, offset, 75, error);
+	fu_byte_array_append_bytes(headerArray, blob_header);
+
+	version = fu_struct_algoltek_product_identity_get_version(headerArray);
+
+	/* len + header = 9 bytes */
+	offset += 9;
+	/* len + productName = 17 bytes */
+	offset += 17;
+	/* len + fwversion = 49 bytes */
+	offset += 49;
+
```

Comment:
this is begging for a rs struct declaration instead

Reply:
Yup, and if you define it there you can also use it in `_validate()`

Reply:
Updated! I used `fu_struct_algoltek_product_identity_validate_bytes( )` in `_validate( )`.